packages/twenty-server/src/engine/core-modules/ask-ai/ask-ai.service.ts

```diff
@@ -0,0 +1,95 @@
+import { Injectable } from '@nestjs/common';
+
+import { ChatOpenAI } from '@langchain/openai';
+import { SqlDatabase } from 'langchain/sql_db';
+import { PromptTemplate } from '@langchain/core/prompts';
+import { RunnableSequence } from '@langchain/core/runnables';
+import { StringOutputParser } from '@langchain/core/output_parsers';
+import { CallbackHandler, Langfuse } from 'langfuse-langchain';
+
+import { AskAIQueryResult } from 'src/engine/core-modules/ask-ai/dtos/ask-ai-query-result.dto';
+import { WorkspaceDataSourceService } from 'src/engine/workspace-datasource/workspace-datasource.service';
+import { WorkspaceQueryRunnerService } from 'src/engine/api/graphql/workspace-query-runner/workspace-query-runner.service';
+
+interface AskAIPromptTemplateInput {
+  schema: string;
+  question: string;
+}
+
+@Injectable()
+export class AskAIService {
+  constructor(
+    private readonly workspaceDataSourceService: WorkspaceDataSourceService,
+    private readonly workspaceQueryRunnerService: WorkspaceQueryRunnerService,
+  ) {}
+
+  async query(
+    userId: string,
+    userEmail: string,
+    workspaceId: string,
+    text: string,
+  ): Promise<AskAIQueryResult> {
+    const workspaceSchemaName =
+      this.workspaceDataSourceService.getSchemaName(workspaceId);
+
+    const workspaceDataSource =
+      await this.workspaceDataSourceService.connectToWorkspaceDataSource(
+        workspaceId,
+      );
+
+    // Does this have unintended side effects? It sets the schema of LangChain's SqlDatabase instance.
+    workspaceDataSource.setOptions({ schema: workspaceSchemaName });
+
+    const db = await SqlDatabase.fromDataSourceParams({
+      appDataSource: workspaceDataSource,
+      sampleRowsInTableInfo: 0,
+    });
+
+    const langfuse = new Langfuse();
+
+    if (!process.env.LANGFUSE_ASK_AI_PROMPT_NAME) throw new Error();
+
+    const promptTemplate =
+      PromptTemplate.fromTemplate<AskAIPromptTemplateInput>(
+        (
+          await langfuse.getPrompt(process.env.LANGFUSE_ASK_AI_PROMPT_NAME)
+        ).getLangchainPrompt(),
+      );
+
+    const model = new ChatOpenAI();
```

Comment:
I would have assumed langchain gave more abstraction and ability to switch more easily to a different model, that's not the case?

Reply:
As we are a self-hosted solution it's important for us to give people choice. If they want to run Mistral it'd be cool if they could (not a MUST but a strong preference if that doesn't cost too much in product quality)